File: flyweight/java/ProductFactory.java
package io.github.androidhot.flyweight.java;

import java.util.EnumMap;
import java.util.Map;

public class ProductFactory {
  private final Map<ProductType, Product> products;

  public ProductFactory() {
    products = new EnumMap<ProductType, Product>(ProductType.class);
  }

  public Product createProduct(ProductType type) {
    Product product = products.get(type);
    if (product == null) {
      switch (type) {
        case PHONE:
          product = new Phone();
          products.put(type, product);
          break;
        case PAD:
          product = new Pad();
          products.put(type, product);
          break;
        case MAC:
          product = new Mac();
          products.put(type, product);
          break;
        case WATCH:
          product = new Watch();
          products.put(type, product);
          break;
        default:
          break;
      }
    }
    return product;
  }
}

File: flyweight/kotlin/ProductFactory.kt
package io.github.androidhot.flyweight.kotlin

import java.util.*

class ProductFactory {
  private val products = EnumMap<ProductType, Product>(ProductType::class.java)

  fun createProduct(type: ProductType): Product {
    var product = products[type]
    if (product == null) {
      product = when(type) {
        ProductType.PHONE -> Phone()
        ProductType.PAD -> Pad()
      }
      products[type] = product
    }
    return product
  }
}
